docs/arch/002-store-responses-in-query-string.md

```diff
@@ -0,0 +1,61 @@
+
+Additionally, the implementation of storing user responses in a session required the creation of a seperate controller `SessionAnswersController`, separate logic for node resolution and updating the flow. This adds extra burden when maintaining or improving the codebase, as new code may need to accommodate both response storage methods.
+
+## Proposal
+
+Add the ability to store user responses using the URL query string. This will become the default method for storing responses and replace use of the URL path. The ability to store responses in a session cookie will remain and require flow to explicitly enable it within configuration (current behaviour).
+
+To implement this change, we could leverage the existing code (controller, node resolution and flow updates) written for storing responses in a session. This is because the underlying data structure storing the responses is the same i.e. a hash keyed by node name. The only additional logic required is to retrieve and update the parameters in the query string instead from the session.
+
+Example:
+- `flow-name/y` => `flow-name/s/question1`
+- `flow-name/y/response1` => `flow-name/s/question2?question1=response1`
+- `flow-name/y/response1/response2` => `flow-name/s/results?question1=response1&question2=question2`
+
+Suggested steps for implementation:
+
+1. Rename `SessionAnswersController` to `FlowController` (as will be used to handle all flows)
+1. Change `use_session` configuration option in flow definition to specify different response store types. e.g. `response_store :session`
+1. Add functionality to store responses in query string
+1. Migrate existing flows incrementally to use query string by specifying config option e.g. `response_store :query_string`
+1. Make query string default response store for all flows
+1. Remove deprecated code for storing responses in URL paths
+
+## Consequences
+
+Our caching effectiveness could be reduced due to different ordering of query parameters. For example, `flow-name/s/results?q1=r1&q2=r2` and `flow-name/s/results?q2=r2&q1=r1` would render the same page, however both hit origin. This is somewhat mitigated by sorting the parameters when generating URLs in Smart Answers, so that users are using the same ordering.
+
+Storing fewer responses in the URL due URL length limit. The length limit of a URL is about 2000 characters (varies between browsers, CDN providers, search engines etc). If the average node name and response is 12 characters, we still can store around 80 responses in the URL using the query string. This is more than enough for our existing flows.
```

Comment:
We currently store responses in the URL structure, so all that the proposal is adding is a few characters per response (i.e. the `q1=` portion). And as you say, that has no bearing on any existing Smart Answer. Might be worth rephrasing it a bit to make it clear this isn't really a consequence:

```suggestion
Smart Answer responses will have longer URLs, as we'll now encode the node name of each question in the URL. Given that the length limit of a URL is about 2000 characters (varies between browsers, CDN providers, search engines etc), and the average node name and response is 12 characters, we can store around 80 responses in the URL using the query string. This is more than enough for our existing flows, the largest of which has X responses.
```

Reply:
Ah thanks for this - your suggestion is far clearer!